### ISSUE TITLE
Introduce lifecycle events

### ISSUE THREAD

**Issue by pandomic** [Steve-Fenton/TypeSpec#20]
Currently we are trying to integrate TypeSpec to test our Angular 4 app. But one big thing that we really need - lifecycle events, like `BeforeScenario` where we can prepare our app and replace dependencies. 

Our current solution - using a custom runner with custom parser with custom feature runner to pass additional params, what is not very handy.

**Comment by Steve-Fenton** [comment]
Depends on Issue #22 as the async handling currently would cause "After..." hooks to run before async tests are complete.

**Comment by Steve-Fenton** [comment]
Test hooks proof of concept is available in master - I'm ready for feedback. At the moment, I have kept it so that the hooks are integrated into the correct triggers, but no data or context is passed.

I have ordered them in the logical order they run.

```
export interface ITestHooks {
    beforeTestRun(): void;
    beforeFeature(): void;
    beforeScenario(): void;
    beforeCondition(): void;

    afterCondition(): void;
    afterScenario(): void;
    afterFeature(): void;
    afterTestRun(): void;
}
```

Console implementation:

```
export class TestHooks implements ITestHooks {
    beforeTestRun(): void {
        console.info('Run Started');
    }

    beforeFeature(): void {
        console.info('|--Feature Started');
    }

    beforeScenario(): void {
        console.info('|--|--Scenario Started');
    }

    beforeCondition(): void {
        console.info('|--|--|--Condition Started');
    }

    afterCondition(): void {
        console.info('|--|--|--Condition Ended');
    }

    afterScenario(): void {
        console.info('|--|--Scenario Ended');
    }

    afterFeature(): void {
        console.info('|--Feature Ended');
    }

    afterTestRun(): void {
        console.info('Run Ended');
    }
}
```

**Comment by Steve-Fenton** [comment]
Please open a new issue / enhancement if you'd like any changes.
